pkg/gds/server.go

```diff
@@ -214,16 +217,85 @@ func (s *Server) Register(ctx context.Context, in *api.RegisterRequest) (out *ap
 		Status:     models.CertificateRequestState_INITIALIZED,
 	}
 
-	if certRequest.Pkcs12Password, certRequest.Pkcs12Signature, err = s.Encrypt(password); err != nil {
-		log.Error().Err(err).Msg("could not encrypt password to store in database")
+	// Create a new Secret Manager
+	sm, err := NewSecretManager(s.conf.Secrets, certRequest.Id)
+
+	// If the application can't access secret manager, we won't be able to keep going
+	// so raise a permissions error and return.
+	if err != nil {
+		log.Error().Err(err).Msg("unable to access secret manager")
 		out.Error = &api.Error{
-			Code:    500,
-			Message: "could not create certificate request password",
+			Code:    403,
+			Message: err.Error(),
+		}
+		return out, nil
+	}
+
+	// Make a new secret of type "password"
+	secret := "password"
+	err = sm.CreateSecret(ctx, secret)
+
+	// If the application can't create a new secret, check to see what the problem is.
+	if err != nil {
+		// If it's a timeout, that means the secret manager API call was bad
+		if errors.Is(err, context.DeadlineExceeded) {
+			log.Error().Err(err).Msg("malformed call to secret manager")
+			out.Error = &api.Error{
+				Code:    400,
+				Message: err.Error(),
+			}
+			return out, nil
+		}
+		// If the secret already exists (which means that the client has previously
+		// created a password), that's okay
+		serr, ok := status.FromError(err)
+		if ok && serr.Code() == codes.AlreadyExists {
+			log.Error().Err(err).Msg("password already exists for client; creating new password")
+		}
+		// If it's a more serious problem, that likely signals a permissions error
+		// and merits a return.
+		log.Error().Err(err).Msg("secret manager error; is GOOGLE_PROJECT_NAME set?")
+		out.Error = &api.Error{
+			Code:    403,
+			Message: err.Error(),
+		}
+		return out, nil
+	}
+
+	// Create a new version for the secret using the password as payload
+	err = sm.AddSecretVersion(ctx, secret, []byte(password))
+	if err != nil {
+		// If it's a timeout, that means the secret manager API call was bad
+		if errors.Is(err, context.DeadlineExceeded) {
+			log.Error().Err(err).Msg("malformed call to secret manager")
+			out.Error = &api.Error{
+				Code:    400,
+				Message: err.Error(),
+			}
+			return out, nil
 		}
+
+		// If the secret is not Not Found, something bad happened.
+		serr, ok := status.FromError(err)
+		if ok && serr.Code() == codes.NotFound {
+			log.Error().Err(err).Msg("secret does not exist, unable to add new secret version")
+			out.Error = &api.Error{
+				Code:    500,
+				Message: err.Error(),
+			}
+			return out, nil
+		}
+
+		log.Error().Err(err).Msg("unable to write new secret version to secret manager")
+		out.Error = &api.Error{
+			Code:    403,
+			Message: err.Error(),
+		}
+		return out, nil
```

Comment:
As above, my review comments have moved a lot of responsibility for error handling to the client. I propose this should be:

```go
if err = sm.AddSecretVersion(ctx, secret, []byte(password)); err != nil {
    log.Error().Err(err).Msg("unable to add secret version on registration")
    out.Error = &api.Error{Code: 500, Message: "internal error during registration, please contact admins"}
    return out, nil
}
```